# Ex 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

https://www.kaggle.com/datasets/giovamata/airlinedelaycauses


In [36]:
!pip install pandas


[notice] A new release of pip available: 22.3.1 -> 23.1.2
[notice] To update, run: pip install --upgrade pip


In [37]:
import numpy as np

In [5]:
import pandas as pd
df = pd.read_csv('DelayedFlights.csv')
df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

Aquí veiem que és un dataset extens (1936758 files × 30 columnes). El dataset consisteix en dades de vols de tot 2018 (des de 3 de gener fins a 13 de desembre).

Però tenim informació truncada a la visualització

In [6]:
# Extendre les columnes de la visualització del dataframe per visualitzar les columnes truncades
# https://builtin.com/data-science/pandas-show-all-columns
pd.set_option('display.max.columns', None)
df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0         2211.0        2225            WN        335  N712SW   
1         1002.0        1000            WN       3231  N772SW   
2          804.0         750            WN        448  N428WN   
3         1959.0        1925            WN       3920  N464WN   
4         2121.0        2110            WN        378  N726SW   
...          ...         ...           ...        ...     ...   
1936753   1617.0        1552            DL       1621  N938DL   
1936754    904.0         749            DL       1631  N3743H   
1936755   1149.0        1010            DL       1631  N909DA   
1936756   1446.0        1437            DL       1639  N646DL   
1936757   1413.0        1418            DL       1641  N908DL   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
0                    128.0           150.0    116.0     -14.0       8.0   
1                    128.0           145.0    113.0       2.0      19.0   
2                     96.0            90.0     76.0      14.0       8.0   
3                     90.0            90.0     77.0      34.0      34.0   
4                    101.0           115.0     87.0      11.0      25.0   
...                    ...             ...      ...       ...       ...   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
0          IAD  TPA       810     4.0      8.0          0                N   
1          IAD  TPA       810     5.0     10.0          0                N   
2          IND  BWI       515     3.0     17.0          0                N   
3          IND  BWI       515     3.0     10.0          0                N   
4          IND  JAX       688     4.0     10.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
1936753    MSP  ATL       906     9.0     18.0          0                N   
1936754    RIC  ATL       481    15.0     34.0          0                N   
1936755    ATL  IAH       689     8.0     32.0          0                N   
1936756    IAD  ATL       533    13.0     13.0          0                N   
1936757    SAT  ATL       874     8.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0               0           NaN           NaN       NaN            NaN   
1               0           NaN           NaN       NaN            NaN   
2               0           NaN           NaN       NaN            NaN   
3               0           2.0           0.0       0.0            0.0   
4               0           NaN           NaN       NaN    

In [7]:
#Per explorar les primeres 50 línies
df.head(50)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0            0  2008      1           3          4   2003.0        1955   
1            1  2008      1           3          4    754.0         735   
2            2  2008      1           3          4    628.0         620   
3            4  2008      1           3          4   1829.0        1755   
4            5  2008      1           3          4   1940.0        1915   
5            6  2008      1           3          4   1937.0        1830   
6           10  2008      1           3          4    706.0         700   
7           11  2008      1           3          4   1644.0        1510   
8           15  2008      1           3          4   1029.0        1020   
9           16  2008      1           3          4   1452.0        1425   
10          17  2008      1           3          4    754.0         745   
11          18  2008      1           3          4   1323.0        1255   
12          19  2008      1           3          4   1416.0        1325   
13          21  2008      1           3          4   1657.0        1625   
14          22  2008      1           3          4   1900.0        1840   
15          23  2008      1           3          4   1039.0        1030   
16          25  2008      1           3          4   1520.0        1455   
17          26  2008      1           3          4   1422.0        1255   
18          27  2008      1           3          4   1954.0        1925   
19          30  2008      1           3          4   2107.0        1945   
20          33  2008      1           3          4   1312.0        1300   
21          34  2008      1           3          4   1449.0        1430   
22          35  2008      1           3          4   1634.0        1555   
23          37  2008      1           3          4   1812.0        1650   
24          38  2008      1           3          4   1127.0        1105   
25          39  2008      1           3          4   1424.0        1355   
26          40  2008      1           3          4   1326.0        1230   
27          41  2008      1           3          4   1749.0        1725   
28          42  2008      1           3          4    726.0         720   
29          43  2008      1           3          4    646.0         640   
30          44  2008      1           3          4   1153.0        1140   
31          45  2008      1           3          4   1528.0        1510   
32          48  2008      1           3          4   1450.0        1435   
33          49  2008      1           3          4   2245.0        1730   
34          52  2008      1           3          4   2025.0        1940   
35          53  2008      1           3          4   1038.0         945   
36          54  2008      1           3          4   1900.0        1850   
37          56  2008      1           3          4    948.0         925   
38          57  2008      1           3          4    646.0         620   
39          58  2008      1           3          4   1110.0        1040   
40          64  2008      1           3          4   2139.0        2130   
41          69  2008      1           3          4   1738.0        1730   
42          70  2008      1           3          4   1813.0        1735   
43          71  2008      1           3          4    802.0         750   
44          74  2008      1           3          4   1734.0        1650   
45          75  2008      1           3          4    712.0         700   
46          76  2008      1           3          4   1318.0        1310   
47          77  2008      1           3          4    958.0         900   
48          78  2008      1           3          4   1859.0        1850   
49          79  2008      1           3          4   1538.0        1445   

    ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0    2211.0        2225            WN        335  N712SW              128.0   
1    1002.0        1

In [8]:
#Per explorar les úñtimes 50 línies
df.tail(50)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1936708     7009454  2008     12          13          6   1259.0        1123   
1936709     7009457  2008     12          13          6   2213.0        2200   
1936710     7009460  2008     12          13          6    847.0         840   
1936711     7009468  2008     12          13          6   1456.0        1122   
1936712     7009469  2008     12          13          6   1645.0        1255   
1936713     7009476  2008     12          13          6   1504.0        1355   
1936714     7009478  2008     12          13          6   1748.0        1722   
1936715     7009479  2008     12          13          6   2109.0        2013   
1936716     7009494  2008     12          13          6   1257.0        1105   
1936717     7009497  2008     12          13          6   1515.0        1302   
1936718     7009498  2008     12          13          6   1319.0        1155   
1936719     7009499  2008     12          13          6   2020.0        1925   
1936720     7009500  2008     12          13          6    945.0         934   
1936721     7009506  2008     12          13          6   1150.0        1128   
1936722     7009513  2008     12          13          6   1355.0        1345   
1936723     7009515  2008     12          13          6   1730.0        1654   
1936724     7009520  2008     12          13          6   1805.0        1718   
1936725     7009548  2008     12          13          6   1618.0        1604   
1936726     7009559  2008     12          13          6   1740.0        1725   
1936727     7009581  2008     12          13          6   1523.0        1515   
1936728     7009590  2008     12          13          6   1512.0        1440   
1936729     7009592  2008     12          13          6    818.0         705   
1936730     7009593  2008     12          13          6   1136.0        1056   
1936731     7009594  2008     12          13          6   1108.0        1100   
1936732     7009603  2008     12          13          6   1443.0        1336   
1936733     7009606  2008     12          13          6   1422.0        1355   
1936734     7009608  2008     12          13          6   1722.0        1646   
1936735     7009617  2008     12          13          6    719.0         710   
1936736     7009627  2008     12          13          6    753.0         735   
1936737     7009628  2008     12          13          6   1132.0        1125   
1936738     7009634  2008     12          13          6   1159.0        1126   
1936739     7009646  2008     12          13          6   1100.0        1045   
1936740     7009652  2008     12          13          6   1200.0        1150   
1936741     7009655  2008     12          13          6    955.0         934   
1936742     7009657  2008     12          13          6   1149.0        1048   
1936743     7009660  2008     12          13          6   1053.0        1000   
1936744     7009664  2008     12          13          6   1251.0        1210   
1936745     7009666  2008     12          13          6   1025.0         950   
1936746     7009671  2008     12          13          6   1429.0        1355   
1936747     7009683  2008     12          13          6   2006.0        1954   
1936748     7009699  2008     12          13          6   1254.0        1221   
1936749     7009701  2008     12          13          6   1528.0        1500   
1936750     7009702  2008     12          13          6   1531.0        1522   
1936751     7009705  2008     12          13          6    921.0         830   
1936752     7009709  2008     12          13          6   1552.0        1520   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757  

In [9]:
# Exploració dels tipus de dades del dataset
print(df.dtypes)

Unnamed: 0             int64
Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


Fins aquí ja es disposa de la informació suficient per explicar les variables del dataset:

### Variables del dataset
<table align="left" text-align="left" cellpadding="20" cellspacing="20" text-align="left">
<tr>
  <th>Variable</th>
  <th>Tipus</th>
  <th>Descripció</th>
</tr>
<tr>
  <td>Year</td>
  <td>int64</td>
  <td>Any (sempre 2008)</td>
</tr>
<tr>
  <td>Month</td>
  <td>int64</td>
  <td>Mes (del 1 al 12)</td>
</tr>
<tr>
  <td>DayofMont</td>
  <td>int64</td>
  <td>Dia del mes (del 1 al 31)</td>
</tr>
<tr>
  <td>DepTime</td>
  <td>float64</td>
  <td>Hora de sortida (hhmm)</td>
</tr>
<tr>
  <td>CRSDepTime</td>
  <td>int64</td>
  <td>Hora programada de sortida (hhmm) (</td>
</tr>
<tr>
  <td>ArrTime</td>
  <td>float64</td>
  <td>Hora d'arribada (hhmm)</td>
</tr>
<tr>
  <td>CSRArrTime</td>
  <td>int64</td>
  <td>Hora programada d'arribada (hhmm)</td>
</tr>
<tr>
  <td>UniqueCarrier</td>
  <td>Object</td>
  <td>Codi IATA de l'operador</td>
</tr>
<tr>
  <td>FlightNum</td>
  <td>int64</td>
  <td>Número de vol</td>
</tr>
<tr>
  <td>TailNum</td>
  <td>Object</td>
  <td>Número de cua de l'avió</td>
</tr>
<tr>
  <td>ActualElapsedTime</td>
  <td>float64</td>
  <td>Temps real de vol (en minuts)</td>
</tr>
<tr>
  <td>CRSElapsedTime</td>
  <td>float64</td>
  <td>Temps programat de vol (en minuts)</td>
</tr>
<tr>
  <td>AirTime</td>
  <td>float64</td>
  <td>Temps enlaire (en minuts)</td>
</tr>
<tr>
  <td>ArrDelay</td>
  <td>float64</td>
  <td>Temps de retard d'arribada (en minuts)</td>
</tr>
<tr>
  <td>DepDelay </td>
  <td>float64</td>
  <td>Temps de retard de sortida (en minuts)</td>
</tr>
<tr>
  <td>Origin</td>
  <td>Object</td>
  <td>Codi IATA d'origen</td>
</tr>
<tr>
  <td>Dest</td>
  <td>Object</td>
  <td>Codi IATA de destinació</td>
</tr>
<tr>
  <td>Distance</td>
  <td>int64</td>
  <td>Distància del vol (en milles)</td>
</tr>
<tr>
  <td>TaxiIn</td>
  <td>float64</td>
  <td>Taxi d'arribada (en minuts)</td>
</tr>
<tr>
  <td>TaxiIn</td>
  <td>float64</td>
  <td>Taxi de sortida (en minuts)</td>
</tr>
<tr>
  <td>Cancelled</td>
  <td>int64</td>
  <td>Vol cancel·lat (booleà, 0 = no; 1 = sí)</td>
</tr>
<tr>
  <td>CancellationCode</td>
  <td>Object</td>
  <td>Codi de cancel·lació</td>
</tr>
<tr>
  <td>Diverted</td>
  <td>int64</td>
  <td>Vol desviat (booleà, 0 = no; 1 = sí)</td>
</tr>
<tr>
  <td>CarrierDelay</td>
  <td>float64</td>
  <td>Retard de la companyia (en minuts)</td>
</tr>
<tr>
  <td>WeatherDelay</td>
  <td>float64</td>
  <td>Retard per l'oratge (en minuts)</td>
</tr>
<tr>
  <td>NASDelay</td>
  <td>float64</td>
  <td>Retard del National Airspace System (en minuts)</td>
</tr>
<tr>
  <td>SecurityDelay</td>
  <td>float64</td>
  <td>Retard per raons de seguretat (en minuts)</td>
</tr>
<tr>
  <td>SecurDelay</td>
  <td>float64</td>
  <td>Retard per l'avió (en minuts)</td>
</tr>
</table>

#### Important

Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.

In [10]:
# Reducció del dataset a 200000 files
df_original = df # Creació d'un back up del dataframe original 
df = df.sample(n=200000) # Selecció aleatoria de 200000 files


In [11]:
df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1399549     4756358  2008      8          22          5   1255.0        1220   
1509423     5306662  2008      9          27          6   1407.0        1400   
1028454     3440942  2008      6          16          1   1408.0        1318   
143994       473171  2008      1           3          4   1014.0        1006   
345660      1086563  2008      2           5          2   1028.0         930   
...             ...   ...    ...         ...        ...      ...         ...   
867111      2936351  2008      5           8          4   2038.0        1945   
185641       610641  2008      2           4          1   1210.0        1155   
682273      2217975  2008      4           6          7   1925.0        1751   
112806       362200  2008      1          12          6   1846.0        1813   
1729785     6444825  2008     11           5          3   1136.0        1125   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
1399549   1827.0        1755            AA       1906  N278AA   
1509423   1510.0        1510            AA       1729  N449AA   
1028454   1701.0        1624            NW       1050  N762NC   
143994    1143.0        1125            9E       5855  87109E   
345660    1649.0        1535            AA       1612  N396AA   
...          ...         ...           ...        ...     ...   
867111    2259.0        2227            B6        598  N705JB   
185641    1403.0        1335            WN        255  N202WN   
682273    2135.0        2023            NW        579  N527US   
112806    2051.0        2035            FL        447  N929AT   
1729785   1458.0        1450            CO         34  N32404   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
1399549              212.0           215.0    182.0      32.0      35.0   
1509423              123.0           130.0     90.0       0.0       7.0   
1028454              113.0           126.0     90.0      37.0      50.0   
143994                89.0            79.0     54.0      18.0       8.0   
345660               261.0           245.0    244.0      74.0      58.0   
...                    ...             ...      ...       ...       ...   
867111               141.0           162.0    133.0      32.0      53.0   
185641               113.0           100.0     68.0      28.0      15.0   
682273               190.0           212.0    168.0      72.0      94.0   
112806                65.0            82.0     52.0      16.0      33.0   
1729785              142.0           145.0    117.0       8.0      11.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
1399549    PHX  ORD      1440     7.0     23.0          0                N   
1509423    DCA  ORD       612     7.0     26.0          0                N   
1028454    MSP  PIT       726     8.0     15.0          0                N   
143994     DTW  ROC       296     7.0     28.0          0                N   
345660     SFO  ORD      1846     6.0     11.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
867111     MCO  HPN       972     4.0      4.0          0                N   
185641     LAS  OAK       407     5.0     40.0          0                N   
682273     MCO  MSP      1310     7.0     15.0          0                N   
112806     MSY  ATL       425     5.0      8.0          0                N   
1729785    DEN  IAH       861    14.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1399549         0           0.0           0.0      26.0            0.0   
1509423         0           NaN           NaN       NaN            NaN   
1028454         0          37.0           0.0       0.0            0.0   
143994          0           8.0           0.0      10.0            0.0   
345660          0          58.0           0.0      16.0    

Ara el dataset és de 200000 rows × 30 columns

Per fer una tria de les variables que són més importants, procedim a fer una anàlisi superficial del dataset i plantegem un possible ús del dataset: conseguir predir quan un vol tindrà retard a partir de certes variables

In [12]:
# Per contabilitzar els valors nuls a cada columna del dataset
print(df.isnull().sum())

Unnamed: 0               0
Year                     0
Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                723
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  1
ActualElapsedTime      858
CRSElapsedTime          19
AirTime                858
ArrDelay               858
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 723
TaxiOut                 40
Cancelled                0
CancellationCode         0
Diverted                 0
CarrierDelay         71307
WeatherDelay         71307
NASDelay             71307
SecurityDelay        71307
LateAircraftDelay    71307
dtype: int64


Les 5 últimes columnes presenten moltes dades NaN (Not a Number), però quin és el percentatge sobre el total de registres de la columna?

In [13]:
# Normalitzar i treure el percentatge de valors NaN o NA de les 5 últimes columnes
df[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].isna().value_counts(normalize=True) * 100


CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay
False         False         False     False          False                64.3465
True          True          True      True           True                 35.6535
Name: proportion, dtype: float64

Amb un total del 35.6805 de valors NaN, aquestes variables poden ser descartades


Per les altres variables que tenen dades nules es pot inferir a simple vista que només un percentatge petit té valors nuls, i per tant es poden acceptar per l'estudi.

<ul>
<li>ArrTime                751</li>
<li>TailNum                  0</li>
<li>ActualElapsedTime      874</li>
<li>CRSElapsedTime          24</li>
<li>AirTime                874</li>
<li>ArrDelay               874</li>
<li>TaxiIn                 751</li>
<li>TaxiOut                 47</li>
</ul>




# Variables del  dataset

Per poder fer una anàlisi de les dades existents i deixar la possibilitat de fer un exercici de predicció amb les variables del dataset, de moment es treuen de l'anàlisi només les darreres 5 columnes, quedant així:
<ul>
<li>Year                      </li>
<li>Month                     </li>
<li>DayofMonth                </li>
<li>DayOfWeek                 </li>
<li>DepTime                   </li>
<li>    CRSDepTime         </li>
<li>ArrTime                </li>
<li>    CRSArrTime                </li>
<li>    UniqueCarrier             </li>
<li>    FlightNum                 </li>
<li>    TailNum                   </li>
<li>ActualElapsedTime      </li>
<li>CRSElapsedTime         </li>
<li>AirTime                </li>
<li>ArrDelay               </li>
<li>DepDelay               </li>
<li>Origin                 </li>
<li>Dest                   </li>
<li>Distance               </li>
    <li>TaxiIn          </li>
<li>    TaxiOut         </li>
<li>Cancelled           </li>
<li>CancellationCode    </li>
<li>Diverted            </li>
    <li><s>CarrierDelay</s></li>
<li><s>    WeatherDelay</s></li>
<li><s>    NASDelay</s></li>
<li><s>    SecurityDelay</s></li>
<li><s>    LateAircraftDelay</s></li>

In [14]:
# Esborrar les últimes 5 línies al dataset i la primera (Unnamed: 0)
df = df.drop(['Unnamed: 0', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'], axis=1)
df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1399549  2008      8          22          5   1255.0        1220   1827.0   
1509423  2008      9          27          6   1407.0        1400   1510.0   
1028454  2008      6          16          1   1408.0        1318   1701.0   
143994   2008      1           3          4   1014.0        1006   1143.0   
345660   2008      2           5          2   1028.0         930   1649.0   
...       ...    ...         ...        ...      ...         ...      ...   
867111   2008      5           8          4   2038.0        1945   2259.0   
185641   2008      2           4          1   1210.0        1155   1403.0   
682273   2008      4           6          7   1925.0        1751   2135.0   
112806   2008      1          12          6   1846.0        1813   2051.0   
1729785  2008     11           5          3   1136.0        1125   1458.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1399549        1755            AA       1906  N278AA              212.0   
1509423        1510            AA       1729  N449AA              123.0   
1028454        1624            NW       1050  N762NC              113.0   
143994         1125            9E       5855  87109E               89.0   
345660         1535            AA       1612  N396AA              261.0   
...             ...           ...        ...     ...                ...   
867111         2227            B6        598  N705JB              141.0   
185641         1335            WN        255  N202WN              113.0   
682273         2023            NW        579  N527US              190.0   
112806         2035            FL        447  N929AT               65.0   
1729785        1450            CO         34  N32404              142.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1399549           215.0    182.0      32.0      35.0    PHX  ORD      1440   
1509423           130.0     90.0       0.0       7.0    DCA  ORD       612   
1028454           126.0     90.0      37.0      50.0    MSP  PIT       726   
143994             79.0     54.0      18.0       8.0    DTW  ROC       296   
345660            245.0    244.0      74.0      58.0    SFO  ORD      1846   
...                 ...      ...       ...       ...    ...  ...       ...   
867111            162.0    133.0      32.0      53.0    MCO  HPN       972   
185641            100.0     68.0      28.0      15.0    LAS  OAK       407   
682273            212.0    168.0      72.0      94.0    MCO  MSP      1310   
112806             82.0     52.0      16.0      33.0    MSY  ATL       425   
1729785           145.0    117.0       8.0      11.0    DEN  IAH       861   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  
1399549     7.0     23.0          0                N         0  
1509423     7.0     26.0          0                N         0  
1028454     8.0     15.0          0                N         0  
143994      7.0     28.0          0                N         0  
345660      6.0     11.0          0                N         0  
...         ...      ...        ...              ...       ...  
867111      4.0      4.0          0                N         0  
185641      5.0     40.0          0                N         0  
682273      7.0     15.0          0                N         0  
112806      5.0      8.0          0                N         0  
1729785    14.0     11.0          0                N         0  

[200000 rows x 24 columns]

In [15]:
# Afegit una columna nova amb index correlatiu
df = df.reset_index()
df

index  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0       1399549  2008      8          22          5   1255.0        1220   
1       1509423  2008      9          27          6   1407.0        1400   
2       1028454  2008      6          16          1   1408.0        1318   
3        143994  2008      1           3          4   1014.0        1006   
4        345660  2008      2           5          2   1028.0         930   
...         ...   ...    ...         ...        ...      ...         ...   
199995   867111  2008      5           8          4   2038.0        1945   
199996   185641  2008      2           4          1   1210.0        1155   
199997   682273  2008      4           6          7   1925.0        1751   
199998   112806  2008      1          12          6   1846.0        1813   
199999  1729785  2008     11           5          3   1136.0        1125   

        ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0        1827.0        1755            AA       1906  N278AA   
1        1510.0        1510            AA       1729  N449AA   
2        1701.0        1624            NW       1050  N762NC   
3        1143.0        1125            9E       5855  87109E   
4        1649.0        1535            AA       1612  N396AA   
...         ...         ...           ...        ...     ...   
199995   2259.0        2227            B6        598  N705JB   
199996   1403.0        1335            WN        255  N202WN   
199997   2135.0        2023            NW        579  N527US   
199998   2051.0        2035            FL        447  N929AT   
199999   1458.0        1450            CO         34  N32404   

        ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin  \
0                   212.0           215.0    182.0      32.0      35.0    PHX   
1                   123.0           130.0     90.0       0.0       7.0    DCA   
2                   113.0           126.0     90.0      37.0      50.0    MSP   
3                    89.0            79.0     54.0      18.0       8.0    DTW   
4                   261.0           245.0    244.0      74.0      58.0    SFO   
...                   ...             ...      ...       ...       ...    ...   
199995              141.0           162.0    133.0      32.0      53.0    MCO   
199996              113.0           100.0     68.0      28.0      15.0    LAS   
199997              190.0           212.0    168.0      72.0      94.0    MCO   
199998               65.0            82.0     52.0      16.0      33.0    MSY   
199999              142.0           145.0    117.0       8.0      11.0    DEN   

       Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  
0       ORD      1440     7.0     23.0          0                N         0  
1       ORD       612     7.0     26.0          0                N         0  
2       PIT       726     8.0     15.0          0                N         0  
3       ROC       296     7.0     28.0          0                N         0  
4       ORD      1846     6.0     11.0          0                N         0  
...     ...       ...     ...      ...        ...              ...       ...  
199995  HPN       972     4.0      4.0          0                N         0  
199996  OAK       407     5.0     40.0          0                N         0  
199997  MSP      1310     7.0     15.0          0                N         0  
199998  ATL       425     5.0      8.0          0                N         0  
199999  IAH       861    14.0     11.0          0                N         0  

[200000 rows x 25 columns]

In [16]:
# Treure l'índex original (ara a segona columna)
df = df.drop(['index'], axis=1)
df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0       2008      8          22          5   1255.0        1220   1827.0   
1       2008      9          27          6   1407.0        1400   1510.0   
2       2008      6          16          1   1408.0        1318   1701.0   
3       2008      1           3          4   1014.0        1006   1143.0   
4       2008      2           5          2   1028.0         930   1649.0   
...      ...    ...         ...        ...      ...         ...      ...   
199995  2008      5           8          4   2038.0        1945   2259.0   
199996  2008      2           4          1   1210.0        1155   1403.0   
199997  2008      4           6          7   1925.0        1751   2135.0   
199998  2008      1          12          6   1846.0        1813   2051.0   
199999  2008     11           5          3   1136.0        1125   1458.0   

        CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0             1755            AA       1906  N278AA              212.0   
1             1510            AA       1729  N449AA              123.0   
2             1624            NW       1050  N762NC              113.0   
3             1125            9E       5855  87109E               89.0   
4             1535            AA       1612  N396AA              261.0   
...            ...           ...        ...     ...                ...   
199995        2227            B6        598  N705JB              141.0   
199996        1335            WN        255  N202WN              113.0   
199997        2023            NW        579  N527US              190.0   
199998        2035            FL        447  N929AT               65.0   
199999        1450            CO         34  N32404              142.0   

        CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
0                215.0    182.0      32.0      35.0    PHX  ORD      1440   
1                130.0     90.0       0.0       7.0    DCA  ORD       612   
2                126.0     90.0      37.0      50.0    MSP  PIT       726   
3                 79.0     54.0      18.0       8.0    DTW  ROC       296   
4                245.0    244.0      74.0      58.0    SFO  ORD      1846   
...                ...      ...       ...       ...    ...  ...       ...   
199995           162.0    133.0      32.0      53.0    MCO  HPN       972   
199996           100.0     68.0      28.0      15.0    LAS  OAK       407   
199997           212.0    168.0      72.0      94.0    MCO  MSP      1310   
199998            82.0     52.0      16.0      33.0    MSY  ATL       425   
199999           145.0    117.0       8.0      11.0    DEN  IAH       861   

        TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  
0          7.0     23.0          0                N         0  
1          7.0     26.0          0                N         0  
2          8.0     15.0          0                N         0  
3          7.0     28.0          0                N         0  
4          6.0     11.0          0                N         0  
...        ...      ...        ...              ...       ...  
199995     4.0      4.0          0                N         0  
199996     5.0     40.0          0                N         0  
199997     7.0     15.0          0                N         0  
199998     5.0      8.0          0                N         0  
199999    14.0     11.0          0                N         0  

[200000 rows x 24 columns]

# Exercici 2
#### Fes un informe complet del dataset:

Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.
Troba quantes dades faltants hi ha per columna.
Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
Fes una taula de les aerolínies amb més endarreriments acumulats.
Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.
Aporta allò que consideris rellevant.

In [17]:
df.describe()

Year          Month     DayofMonth      DayOfWeek        DepTime  \
count  200000.0  200000.000000  200000.000000  200000.000000  200000.000000   
mean     2008.0       6.105360      15.768305       3.981635    1518.936885   
std         0.0       3.481238       8.747010       1.996486     450.425703   
min      2008.0       1.000000       1.000000       1.000000       1.000000   
25%      2008.0       3.000000       8.000000       2.000000    1203.000000   
50%      2008.0       6.000000      16.000000       4.000000    1545.000000   
75%      2008.0       9.000000      23.000000       6.000000    1900.000000   
max      2008.0      12.000000      31.000000       7.000000    2400.000000   

          CRSDepTime        ArrTime     CRSArrTime      FlightNum  \
count  200000.000000  199277.000000  200000.000000  200000.000000   
mean     1467.812910    1609.044581    1632.974630    2178.264330   
std       425.006497     548.841451     465.734234    1942.343779   
min         1.000000       1.000000       0.000000       1.000000   
25%      1135.000000    1315.000000    1325.000000     607.000000   
50%      1510.000000    1714.000000    1705.000000    1537.000000   
75%      1815.000000    2030.000000    2014.000000    3420.000000   
max      2359.000000    2400.000000    2400.000000    9740.000000   

       ActualElapsedTime  CRSElapsedTime        AirTime       ArrDelay  \
count      199142.000000   199981.000000  199142.000000  199142.000000   
mean          133.254442      134.276761     108.238172      42.130560   
std            71.981165       71.375102      68.638664      56.392842   
min            16.000000      -10.000000       0.000000     -61.000000   
25%            80.000000       81.000000      58.000000       9.000000   
50%           116.000000      116.000000      90.000000      24.000000   
75%           165.000000      165.000000     137.000000      56.000000   
max           750.000000      660.000000     651.000000    1392.000000   

            DepDelay      Distance         TaxiIn        TaxiOut  \
count  200000.000000  200000.00000  199277.000000  199960.000000   
mean       43.115775     765.72414       6.818910      18.215633   
std        53.014417     575.26704       5.305676      14.202729   
min         6.000000      30.00000       0.000000       0.000000   
25%        12.000000     338.00000       4.000000      10.000000   
50%        24.000000     606.00000       6.000000      14.000000   
75%        53.000000     998.00000       8.000000      21.000000   
max      1369.000000    4962.00000     186.000000     379.000000   

           Cancelled      Diverted  
count  200000.000000  200000.00000  
mean        0.000300       0.00399  
std         0.017318       0.06304  
min         0.000000       0.00000  
25%         0.000000       0.00000  
50%         0.000000       0.00000  
75%         0.000000       0.00000  
max         1.000000       1.00000

In [18]:
# Per visualitzat millor el resultat, arrodonim els valors
round(df.describe(),2)

Year      Month  DayofMonth  DayOfWeek    DepTime  CRSDepTime  \
count  200000.0  200000.00   200000.00  200000.00  200000.00   200000.00   
mean     2008.0       6.11       15.77       3.98    1518.94     1467.81   
std         0.0       3.48        8.75       2.00     450.43      425.01   
min      2008.0       1.00        1.00       1.00       1.00        1.00   
25%      2008.0       3.00        8.00       2.00    1203.00     1135.00   
50%      2008.0       6.00       16.00       4.00    1545.00     1510.00   
75%      2008.0       9.00       23.00       6.00    1900.00     1815.00   
max      2008.0      12.00       31.00       7.00    2400.00     2359.00   

         ArrTime  CRSArrTime  FlightNum  ActualElapsedTime  CRSElapsedTime  \
count  199277.00   200000.00  200000.00          199142.00       199981.00   
mean     1609.04     1632.97    2178.26             133.25          134.28   
std       548.84      465.73    1942.34              71.98           71.38   
min         1.00        0.00       1.00              16.00          -10.00   
25%      1315.00     1325.00     607.00              80.00           81.00   
50%      1714.00     1705.00    1537.00             116.00          116.00   
75%      2030.00     2014.00    3420.00             165.00          165.00   
max      2400.00     2400.00    9740.00             750.00          660.00   

         AirTime   ArrDelay   DepDelay   Distance     TaxiIn    TaxiOut  \
count  199142.00  199142.00  200000.00  200000.00  199277.00  199960.00   
mean      108.24      42.13      43.12     765.72       6.82      18.22   
std        68.64      56.39      53.01     575.27       5.31      14.20   
min         0.00     -61.00       6.00      30.00       0.00       0.00   
25%        58.00       9.00      12.00     338.00       4.00      10.00   
50%        90.00      24.00      24.00     606.00       6.00      14.00   
75%       137.00      56.00      53.00     998.00       8.00      21.00   
max       651.00    1392.00    1369.00    4962.00     186.00     379.00   

       Cancelled   Diverted  
count  200000.00  200000.00  
mean        0.00       0.00  
std         0.02       0.06  
min         0.00       0.00  
25%         0.00       0.00  
50%         0.00       0.00  
75%         0.00       0.00  
max         1.00       1.00

Aquí veiem que els vols més llargs són de 4962.00 milles i de 649.00 minuts.<br/>
El més endarrerit és de 2461.00 minuts<br/>
El vol més curts és de 24 milles<br/>
La distància mitjana és de 764.60 milles<br/>


#### Troba quantes dades faltants hi ha per columna. 

In [19]:
# Per contabilitzar els valors nuls a cada columna del dataset
print(df.isnull().sum())

Year                   0
Month                  0
DayofMonth             0
DayOfWeek              0
DepTime                0
CRSDepTime             0
ArrTime              723
CRSArrTime             0
UniqueCarrier          0
FlightNum              0
TailNum                1
ActualElapsedTime    858
CRSElapsedTime        19
AirTime              858
ArrDelay             858
DepDelay               0
Origin                 0
Dest                   0
Distance               0
TaxiIn               723
TaxiOut               40
Cancelled              0
CancellationCode       0
Diverted               0
dtype: int64


Per tractar els NaN, anem columna per columna.

He plantejat, en el cas de ArrTime, que es pot deduir sumant a Departure Time els valors de Actual Elapsed Time. Tanmateix, s'ha de convertir del format hora (hhmm) a minuts i després altre cop a format hora. Massa complicat i no sabem si la nova hora resultant és hora local o no, i per tant considero millor eliminar la columna sencera.

In [20]:
df1 = df.drop(['ArrTime'], axis=1)
df1

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled CancellationCode  Diverted  
0               0                N         0  
1               0                N         0  
2               0                N         0  
3               0                N         0  
4               0                N         0  
...           ...              ...       ...  
199995          0                N         0  
199996          0                N         0  
199997          0                N         0  
199998          0                N         0  
199999          0                N         0  

[200000 rows x 23 columns]

In [21]:
df1.isnull().sum()

Year                   0
Month                  0
DayofMonth             0
DayOfWeek              0
DepTime                0
CRSDepTime             0
CRSArrTime             0
UniqueCarrier          0
FlightNum              0
TailNum                1
ActualElapsedTime    858
CRSElapsedTime        19
AirTime              858
ArrDelay             858
DepDelay               0
Origin                 0
Dest                   0
Distance               0
TaxiIn               723
TaxiOut               40
Cancelled              0
CancellationCode       0
Diverted               0
dtype: int64

Tenim 3 valors iguals. Per què?<br/>
ActualElapsedTime = 858<br/>
AirTime = 858<br/>
ArrDelay = 858<br/>

Anem a explorar les dades per veure les relacions entre les variables:

In [22]:
# Explorar quants vols han estat cancel·lats o desviats (typus de dada: int64, booleà 1 o 0)
df1[['Cancelled', 'Diverted']].value_counts()

Cancelled  Diverted
0          0           199142
           1              798
1          0               60
Name: count, dtype: int64

In [23]:
# Explorar quants vols han estat cancel·lats o desviats (int64, format hora)
df1[['ActualElapsedTime']].value_counts()

ActualElapsedTime
80.0                 1749
75.0                 1696
81.0                 1689
78.0                 1678
77.0                 1675
                     ... 
537.0                   1
536.0                   1
535.0                   1
534.0                   1
750.0                   1
Name: count, Length: 579, dtype: int64

D'aquí deduim que la relació de 'ActualElapsedTime' es pot explicar perquè els vols 'Cancelled' i 'Diverted' no tenen temps real de vol. Per comprovar-ho, treiem els vols 'Cancelled' i 'Diverted' primer i veiem el resultat.

In [24]:
df2 = df1 #fem un nou dataframe per no perdre les dades en cas d'error
df2

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled CancellationCode  Diverted  
0               0                N         0  
1               0                N         0  
2               0                N         0  
3               0                N         0  
4               0                N         0  
...           ...              ...       ...  
199995          0                N         0  
199996          0                N         0  
199997          0                N         0  
199998          0                N         0  
199999          0                N         0  

[200000 rows x 23 columns]

In [25]:
# Creació de variable per guardar els vols cancel·lats
vols_cancellats = df2[df2['Cancelled']==1]
len(vols_cancellats)

60

In [26]:
# Extracció dels registres on el vol està cancel·lat 
df2 = df2[df2['Cancelled']!=1]
df2

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled CancellationCode  Diverted  
0               0                N         0  
1               0                N         0  
2               0                N         0  
3               0                N         0  
4               0                N         0  
...           ...              ...       ...  
199995          0                N         0  
199996          0                N         0  
199997          0                N         0  
199998          0                N         0  
199999          0                N         0  

[199940 rows x 23 columns]

In [27]:
# Comprovem quins valors nuls queden al dataframe
df2.isnull().sum()

Year                   0
Month                  0
DayofMonth             0
DayOfWeek              0
DepTime                0
CRSDepTime             0
CRSArrTime             0
UniqueCarrier          0
FlightNum              0
TailNum                1
ActualElapsedTime    798
CRSElapsedTime        19
AirTime              798
ArrDelay             798
DepDelay               0
Origin                 0
Dest                   0
Distance               0
TaxiIn               663
TaxiOut                0
Cancelled              0
CancellationCode       0
Diverted               0
dtype: int64

Veiem que efectivament s'han restat 68 vols en totes les columnes. Per tant, els valors nuls d'altre columnes venen donants per 'Cancelled´. Fem lo mateix amb 'Diverted':

In [28]:
# Creació de variable per guardar els vols desviats
vols_desviats = df2[df2['Diverted']==1]
print(len(vols_desviats))

798


798 és justament els vols amb valors nulls d'altres columnes.

In [29]:
# Extracció dels registres on el vol està desviat
df2 = df2[df2['Diverted']!=1]
df2


Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled CancellationCode  Diverted  
0               0                N         0  
1               0                N         0  
2               0                N         0  
3               0                N         0  
4               0                N         0  
...           ...              ...       ...  
199995          0                N         0  
199996          0                N         0  
199997          0                N         0  
199998          0                N         0  
199999          0                N         0  

[199142 rows x 23 columns]

In [30]:
# Comprovem quins valors nuls queden al dataframe
df2.isnull().sum()

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
dtype: int64

Aquí veiem que amb les columnes 'Cancelled' i 'Diverted' hem tret tots els valors nuls del dataset. Continuem eliminant la columna 'CancellationCode', que s'ha quedat buida de sentit

In [31]:
# Treure columnca 'CancellationCode'
df2 = df2.drop('CancellationCode', axis=1)
print(df2)

        Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElap

#### Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...). 

In [110]:
df3 = df2 #Nou df per no perdre dades anteriors

# Creació de nova columna amb la velocitat mitjana en milles per hora, dividint la distància pels temps real de vol i multiplicant per 60
df3['AvgSpeed'] =  (df3['Distance'] / df3['ActualElapsedTime']) * 60


In [113]:
df3

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled  Diverted    AvgSpeed  
0               0         0  407.547170  
1               0         0  298.536585  
2               0         0  385.486726  
3               0         0  199.550562  
4               0         0  424.367816  
...           ...       ...         ...  
199995          0         0  413.617021  
199996          0         0  216.106195  
199997          0         0  413.684211  
199998          0         0  392.307692  
199999          0         0  363.802817  

[199142 rows x 23 columns]

In [33]:
# Arrodonir els valors a dos decimals
df3['AvgSpeed'] = df3['AvgSpeed'].round(2)
df3

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled  Diverted  AvgSpeed  
0               0         0    407.55  
1               0         0    298.54  
2               0         0    385.49  
3               0         0    199.55  
4               0         0    424.37  
...           ...       ...       ...  
199995          0         0    413.62  
199996          0         0    216.11  
199997          0         0    413.68  
199998          0         0    392.31  
199999          0         0    363.80  

[199142 rows x 23 columns]

In [114]:
# Per saber quins vols han arribat tard en una nova columna amb booleans (1 = Yes, 0 = No)
df3['DelayedFlight'] = np.where(df3['ArrDelay'] > 0, '1', '0')
df3

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled  Diverted    AvgSpeed DelayedFlight  
0               0         0  407.547170             1  
1               0         0  298.536585             0  
2               0         0  385.486726             1  
3               0         0  199.550562             1  
4               0         0  424.367816             1  
...           ...       ...         ...           ...  
199995          0         0  413.617021             1  
199996          0         0  216.106195             1  
199997          0         0  413.684211             1  
199998          0         0  392.307692             1  
199999          0         0  363.802817             1  

[199142 rows x 24 columns]

#### Fes una taula de les aerolínies amb més endarreriments acumulats.

In [115]:
unique_carrier_delays = pd.DataFrame(df3.groupby('UniqueCarrier')['ArrDelay'].sum().sort_values(ascending=False))
unique_carrier_delays.head(5)

ArrDelay
UniqueCarrier           
WN             1156891.0
AA              909548.0
UA              714151.0
MQ              657622.0
OO              615213.0

El top 5 d'aerolínies amb més retards són les següents:<br/>
WN - Southwest Airlines<br/>
AA - American Airlines<br/>
UA - United Airlines<br/>
MQ - Envoy Air (American Eagle)<br/>
OO - SkyWest Airlines<br/>


####  Quins són els vols més llargs?   

In [66]:
# Els més llargs de distància
longest_flights = df3[['FlightNum', 'Distance']].sort_values(by='Distance', ascending=False)
longest_flights.head(5)

FlightNum  Distance
20108          14      4962
170476         14      4962
1429           14      4962
187783         15      4962
93867          15      4962

In [69]:
# Els de més duració. Prenc 'ActualElapsedTime' com referència
timest_flights = df3[['FlightNum', 'AirTime', 'ActualElapsedTime']].sort_values(by='ActualElapsedTime', ascending=False)
timest_flights.head(5)

FlightNum  AirTime  ActualElapsedTime
99558          15    597.0              750.0
169187         15    594.0              724.0
120944         15    641.0              693.0
16745          15    651.0              679.0
187783         15    641.0              667.0

#### I els més endarrerits?

In [71]:
# Els de més duració. Prenc 'ActualElapsedTime' com referència
timest_flights = df3[['FlightNum', 'AirTime', 'ActualElapsedTime']].sort_values(by='ActualElapsedTime', ascending=True)
timest_flights.head(5)

FlightNum  AirTime  ActualElapsedTime
139843         65      9.0               16.0
153820       1030      7.0               17.0
177681         65      8.0               17.0
81376          64     10.0               17.0
12249          64      9.0               19.0

#### Busca les rutes més llargues

In [81]:
# Els més llargs de distància, incloent-hi 'Origin' => 'Dest' & 'Dest' => 'Origin' com a mateixa ruta
longest_routes = df3[['Origin', 'Dest', 'Distance']].sort_values(by='Distance', ascending=False)
longest_routes = longest_routes.drop_duplicates(subset=['Distance']) # Eliminar rutes Origen-Destí duplicades o amb la mateixa distància
longest_routes.head(5)

Origin Dest  Distance
20108     HNL  EWR      4962
92127     ATL  HNL      4502
105633    HNL  ORD      4243
30371     KOA  ORD      4213
149373    ORD  OGG      4184

#### i les que acumulen més retards.

In [98]:
# Els més llargs de distància
timest_routes = pd.DataFrame(df3.groupby(['Origin', 'Dest'])['DelayedFlight'].sum())
timest_routes = timest_routes.sort_values(by='DelayedFlight', ascending=False)

timest_routes.head(5)

DelayedFlight
Origin Dest                                                   
HNL    OGG   1111111111111111111111111111111111111111111111...
OGG    HNL   1111111111111111111111111111111111111111111111...
HNL    LIH   1111111111111111111111111111111111111111111111...
KOA    HNL   1111111111111111111111111111111111111111111111...
LIH    HNL   1111111111111111111111111111111111111111111111...

In [102]:
# Com que concatena els 1, comprovem el tipus de dada que és 'DelayedFlight', ja que sembla que no és integer
print(df3['DelayedFlight'].dtypes)

object


In [104]:
# Convertim la columna a integer
df3['DelayedFlight'] = df3['DelayedFlight'].astype(int)


In [105]:
# Repetim l'operació per trobar les rutes amb més retards
timest_routes = pd.DataFrame(df3.groupby(['Origin', 'Dest'])['DelayedFlight'].sum())
timest_routes = timest_routes.sort_values(by='DelayedFlight', ascending=False)

timest_routes.head(5)

DelayedFlight
Origin Dest               
LAX    SFO             478
SFO    LAX             397
ORD    LGA             373
ATL    LGA             363
DAL    HOU             337

En conclusió:
LAX a SFO amb 478 vols retardats.<br/>
SFO a LAX amb 397.<br/>
ORD a LGA amb 373.<br/>
ATL a LGA amb 363.<br/>
DAL a HOU amb 337.<br/>



Per últim, afegim el percentatge total de vols retardats en 2008

In [122]:
# Normalitzar i treure el percentatge de valors NaN o NA de les 5 últimes columnes
ptg_delayed_flights = df3[['DelayedFlight']].value_counts(normalize=True) * 100
ptg_delayed_flights.round(2)

DelayedFlight
1                89.37
0                10.63
Name: proportion, dtype: float64

#### Podem concluir que totes aquestes dades han estat analitzades amb un percentatge de retards del 10.63%

# Exercici 3
#### Exporta el dataset net i amb les noves columnes a Excel.

In [118]:
df3 #comprovem si està tot ok abans d'exportar

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  CRSArrTime  \
0       2008      8          22          5   1255.0        1220        1755   
1       2008      9          27          6   1407.0        1400        1510   
2       2008      6          16          1   1408.0        1318        1624   
3       2008      1           3          4   1014.0        1006        1125   
4       2008      2           5          2   1028.0         930        1535   
...      ...    ...         ...        ...      ...         ...         ...   
199995  2008      5           8          4   2038.0        1945        2227   
199996  2008      2           4          1   1210.0        1155        1335   
199997  2008      4           6          7   1925.0        1751        2023   
199998  2008      1          12          6   1846.0        1813        2035   
199999  2008     11           5          3   1136.0        1125        1450   

       UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0                 AA       1906  N278AA              212.0           215.0   
1                 AA       1729  N449AA              123.0           130.0   
2                 NW       1050  N762NC              113.0           126.0   
3                 9E       5855  87109E               89.0            79.0   
4                 AA       1612  N396AA              261.0           245.0   
...              ...        ...     ...                ...             ...   
199995            B6        598  N705JB              141.0           162.0   
199996            WN        255  N202WN              113.0           100.0   
199997            NW        579  N527US              190.0           212.0   
199998            FL        447  N929AT               65.0            82.0   
199999            CO         34  N32404              142.0           145.0   

        AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0         182.0      32.0      35.0    PHX  ORD      1440     7.0     23.0   
1          90.0       0.0       7.0    DCA  ORD       612     7.0     26.0   
2          90.0      37.0      50.0    MSP  PIT       726     8.0     15.0   
3          54.0      18.0       8.0    DTW  ROC       296     7.0     28.0   
4         244.0      74.0      58.0    SFO  ORD      1846     6.0     11.0   
...         ...       ...       ...    ...  ...       ...     ...      ...   
199995    133.0      32.0      53.0    MCO  HPN       972     4.0      4.0   
199996     68.0      28.0      15.0    LAS  OAK       407     5.0     40.0   
199997    168.0      72.0      94.0    MCO  MSP      1310     7.0     15.0   
199998     52.0      16.0      33.0    MSY  ATL       425     5.0      8.0   
199999    117.0       8.0      11.0    DEN  IAH       861    14.0     11.0   

        Cancelled  Diverted    AvgSpeed DelayedFlight  
0               0         0  407.547170             1  
1               0         0  298.536585             0  
2               0         0  385.486726             1  
3               0         0  199.550562             1  
4               0         0  424.367816             1  
...           ...       ...         ...           ...  
199995          0         0  413.617021             1  
199996          0         0  216.106195             1  
199997          0         0  413.684211             1  
199998          0         0  392.307692             1  
199999          0         0  363.802817             1  

[199142 rows x 24 columns]

In [123]:
# Hi queden valors nuls?
df3.isnull().sum()

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
Diverted             0
AvgSpeed             0
DelayedFlight        0
dtype: int64

Procedim a exportar

In [108]:
!pip install openpyxl

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 250.0/250.0 kB 3.9 MB/s eta 0:00:00a 0:00:01

[notice] A new release of pip available: 22.3.1 -> 23.1.2
[notice] To update, run: pip install --upgrade pip


In [109]:
df3.to_excel('DelayedFlightStats.xlsx')